code/datums/components/transfers.dm

```diff
@@ -0,0 +1,227 @@
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.transfer_proc = transfer
+	src.filter = filter
+	src.filter_proc = filter_proc
+	src.filter_link_proc = filter_link_proc
+
+/datum/component/transfer_input/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_INCOMING_TRANSFER, .proc/handle_incoming)
+	RegisterSignal(parent, COMSIG_CAN_LINK, .proc/handle_incoming_link)
+	RegisterSignal(parent, COMSIG_ATTACKBY, .proc/handle_attackby)
+
+/datum/component/transfer_input/proc/handle_incoming_link(comsig_target, obj/other)
+	return !filter_link_proc || call(parent, filter_link_proc)(other)
+
+/datum/component/transfer_input/proc/handle_incoming(comsig_target, obj/item/incoming)
+	if (!isitem(incoming))
+		return
+
+	if (filter && !(incoming.type in filter))
+		return
+
+	if(!filter_proc || call(parent, filter_proc)(incoming))
+		incoming.set_loc(parent)
+		if (transfer_proc)
+			call(parent, transfer_proc)(incoming)
+		return TRUE
+
+#define CONTAINER_CHOICE_PLACE "Place container inside"
+#define CONTAINER_CHOICE_DUMP  "Dump its contents inside"
+
+/datum/component/transfer_input/proc/handle_attackby(comsig_target, obj/item/incoming, mob/attacker)
+	if(!isitem(incoming))
+		return
+
+	if (istype(incoming, /obj/item/deconstructor) || isgrab(incoming))
+		return
+
+	if(cant_do_shit(attacker))
+		return
+
+	if (istype(incoming, /obj/item/magtractor))
+		var/obj/item/magtractor/M = incoming
+		if (M.holding)
+			incoming = M.holding
+
+	if(incoming.cant_drop)
+		boutput(attacker, "<span class='alert'>You can't put that in [parent] when it's attached to you!</span>")
+		return
+
+	if ((istype(incoming, /obj/item/storage) || istype(incoming, /obj/item/satchel) || istype(incoming, /obj/item/ore_scoop)) && length(incoming.contents))
+		var/action
+		if(!filter_proc || call(parent, filter_proc)(incoming))
+			action = input(attacker, "What do you want to do with [incoming]?") as null|anything in list(CONTAINER_CHOICE_PLACE, CONTAINER_CHOICE_DUMP)
+		else
+			action = CONTAINER_CHOICE_DUMP
+
+		if (!action)
+			return
+
+		if (!in_interact_range(parent, attacker))
+			boutput(attacker, "<span class='alert'>You need to be closer to [parent] to do that.</span>")
+			return
+
+		if (action == CONTAINER_CHOICE_DUMP)
+			if (istype(incoming, /obj/item/ore_scoop))
+				var/obj/item/ore_scoop/scoop = incoming
+				incoming = scoop.satchel
+			for(var/obj/item/I in incoming)
+				if (SEND_SIGNAL(parent, COMSIG_INCOMING_TRANSFER, I))
+					attacker.u_equip(I)
+					I.dropped()
+			attacker.visible_message("<span class='notice'>[attacker] dumps out [incoming] into [parent].</span>")
+			return TRUE
+
+	if (SEND_SIGNAL(parent, COMSIG_INCOMING_TRANSFER, incoming))
+		attacker.visible_message("<span class='notice'>[attacker] places [incoming] into [parent]</span>")
+		attacker.u_equip(incoming)
+		if(isitem(incoming))
+			var/obj/item/I = incoming
+			I.dropped()
```

Comment:
Doesn't need this u_equip? God I hate the unequipping/dropping/equipping mess of a code.

Reply:
u_equip call required, the item does not leave the interface when it is removed.